### PR TITLE
Vectorize keypoint batching loop for Pose loss calculation

### PR DESCRIPTION
**Pull request by ahmet-f-gumustas**
## Summary
- Replace the sequential `for` loop in `_keypoint_select` with fully vectorized tensor operations using `scatter_add_` and advanced indexing
- Resolves the TODO comment at loss.py:716

## Details
The previous implementation iterated over each batch index to fill `batched_keypoints`:

```python
for i in range(batch_size):
    keypoints_i = keypoints[batch_idx == i]
    batched_keypoints[i, : keypoints_i.shape[0]] = keypoints_i

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
⚡ This PR vectorizes keypoint batching in `ultralytics/utils/loss.py`, replacing a per-batch Python loop with indexed tensor operations for a cleaner and potentially faster pose loss path.

### 📊 Key Changes
- Replaces the loop that iterated over each batch element to populate `batched_keypoints`.
- Introduces a vectorized indexing approach using `batch_idx`, `scatter_add_`, cumulative offsets, and `within_idx`.
- Preserves the existing output structure and downstream target keypoint selection flow.
- Updates logic specifically in `_select_target_keypoints()`, which is part of the pose/keypoint loss pipeline.

### 🎯 Purpose & Impact
- Improves efficiency by reducing Python-side looping in a performance-sensitive loss computation path. 🚀
- Makes batching of pose keypoints more scalable for larger batch sizes or denser annotations. 📈
- Keeps functionality focused and minimal, lowering integration risk while improving implementation quality. ✅
- Most relevant to YOLO pose training workloads, where repeated loss computations can benefit from vectorization. 🧠

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ahmet-f-gumustas, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with the review process—an engineer will assist you shortly. Please review the checklist below to help ensure a smooth integration of your update to the pose loss pipeline. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/loss.py [hunk 1]
```diff
@@ -713,11 +713,13 @@ def _select_target_keypoints(
...
-        # Fill batched_keypoints with keypoints based on batch_idx
-        for i in range(batch_size):
-            keypoints_i = keypoints[batch_idx == i]
-            batched_keypoints[i, : keypoints_i.shape[0]] = keypoints_i
+        # Vectorized fill: compute within-batch position for each keypoint using cumulative offsets
+        batch_idx_long = batch_idx.long()
+        offsets = torch.zeros(batch_size + 1, dtype=torch.long, device=keypoints.device)
+        offsets.scatter_add_(0, batch_idx_long + 1, torch.ones_like(batch_idx_long))
+        offsets = offsets.cumsum(0)
+        within_idx = torch.arange(len(batch_idx), device=keypoints.device) - offsets[batch_idx_long]
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/loss.py:721]: ⚠️ **HIGH**: This indexing only produces correct `within_idx` values when `batch_idx` is already grouped by batch. If annotations are interleaved across images, the first occurrence of a later batch will get a nonzero slot and can leave gaps or even exceed `max_kpts`, whereas the removed loop always packed each batch from index 0. Unless contiguity is guaranteed here, this changes behavior for valid input orderings.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The vectorization looks good conceptually, but it currently relies on `batch_idx` being sorted/grouped by image. That assumption was not required before, so this introduces a correctness risk in `_select_target_keypoints()` for interleaved targets.

💬 Posted 1 inline comment

File: ultralytics/utils/loss.py (hunk 1 above)
Comment (ahmet-f-gumustas) [reply ultralytics/utils/loss.py:721]: batch_idx is always contiguous/grouped by batch index. It is constructed in YOLODataset.collate_fn ([dataset.py:304-307](vscode-webview://1o22g31o860dmhsq9ome0ub9p582l2bfvc5i93h1b7fjsj2372bb/ultralytics/data/dataset.py#L304-L307)) by iterating over each image in order and concatenating their annotations sequentially. Image 0's annotations come first, then image 1's, etc. So the grouping assumption holds for all valid inputs — interleaved batch indices cannot occur through the standard data pipeline.

**Review by lmycross — COMMENTED** [review]
lgtm

**Review by Laughing-q — APPROVED** [review]

**Comment by glenn-jocher** [comment]
Thanks, this was a clean, scoped optimization to `_select_target_keypoints()` and is now merged in `1282f03`; the same vectorization pattern was also carried into `#23966`.
